### DISCUSSION TITLE
Vapore pressure calculations for aqueous system with PengRobinsonPhreeqc at low temp (<100 C) and ambient pressure

### DISCUSSION THREAD

**Discussion by avdudchenko** [reaktoro/reaktoro#398] [Q&A]
### Description

I am trying to configure reaktoro to compute the vapor pressure of an aqueous system containing salts. When I use a simple setup adding in "H2O(g)" phase to get its partial pressure, I find that the final equilibrium state does not make logical sense (most of water turns in the gas phase, effectively forming sludge and gas vapor). 

The output for the equilibrium state is shown in below. Although, I get nearly the same partial pressure as from PhreeqC (1969 Pa from reaktoro vs 2351 Pa for phreeqc), the difference is likely a result of the final state we observe here (the high salinity results in significant depression). 

I wanted to see if I was missing something. I tried adding constraints to vapor volume and amounts of phase but either got a similar result or the equilibrium failed to solve. 

This feels like it also could be related to the discussion in [285 ](https://github.com/reaktoro/reaktoro/discussions/285) but it sounded like the new ActivityModelPengRobinsonPhreeqc fixed the issue. Not sure? Any advice/recommendations would be appreciated.  

```
+-----------------+------------+------+
| Property        |      Value | Unit |
+-----------------+------------+------+
| Temperature     |   293.1500 |    K |
| Pressure        |     1.0000 |  bar |
| Charge:         | 2.3661e-12 |  mol |
| Element Amount: |            |      |
| :: H            | 1.1103e+02 |  mol |
| :: C            | 8.1945e-03 |  mol |
| :: O            | 5.5534e+01 |  mol |
| :: Na           | 4.8717e-02 |  mol |
| :: S            | 1.0525e-05 |  mol |
| :: Cl           | 4.8207e-02 |  mol |
| :: Ca           | 3.7427e-03 |  mol |
| Species Amount: |            |      |
| :: H+           | 4.0815e-10 |  mol |
| :: H2O          | 6.9787e-01 |  mol |
| :: CO3-2        | 1.7675e-04 |  mol |
| :: CO2          | 3.9700e-04 |  mol |
| :: Ca+2         | 3.7427e-03 |  mol |
| :: Cl-          | 4.8207e-02 |  mol |
| :: HCO3-        | 7.6207e-03 |  mol |
| :: SO4-2        | 1.0525e-05 |  mol |
| :: HSO4-        | 1.2003e-12 |  mol |
| :: Na+          | 4.8717e-02 |  mol |
| :: OH-          | 2.7483e-09 |  mol |
| :: H2O(g)       | 5.4812e+01 |  mol |
+-----------------+------------+------+
```

### Example

```py
import reaktoro as rkt

db = rkt.PhreeqcDatabase("pitzer.dat")
# ph adjusted and charge neutralized 
feed_comp = {
    "H+": 1.9558980691081993e-08,
    "H2O": 55.50929780738274,
    "CO3-2": 4.905954807916131e-05,
    "CO2": 0.00026983065832876047,
    "Ca+2": 0.0037427017316233346,
    "Cl-": 0.048207460278976974,
    "HCO3-": 0.007875574268951177,
    "SO4-2": 1.0524665304507918e-05,
    "HSO4-": 5.9841513528777e-12,
    "Na+": 0.04871732291003011,
    "OH-": 5.429515778317798e-07,
}

"""solve with added gas phase"""
solution = rkt.AqueousPhase(rkt.speciate(" ".join(list(feed_comp.keys()))))
solution.set(rkt.ActivityModelPitzer())
gas = rkt.GaseousPhase("H2O(g)")
gas.set(rkt.ActivityModelPengRobinsonPhreeqc())
system_with_gas = rkt.ChemicalSystem(db, solution, gas)
state_with_gas = rkt.ChemicalState(system_with_gas)

for ion, values in feed_comp.items():
    state_with_gas.set(ion, values, "mol")
state_with_gas.temperature(293.15, "kelvin")
state_with_gas.pressure(1, "bar")

rkt.equilibrate(state_with_gas)
vapor_pressure = state_with_gas.props().speciesActivity("H2O(g)") * 101325
print(state_with_gas)
print(
    "expected vapor pressure",
    2351.213686124685,
    " (Pa), reaktoro vapor pressure",
    vapor_pressure,
    "(Pa)",
)
```

### What is your operating system?

None

**Comment by allanleal** [comment]
Hi @avdudchenko - the cubic equations of state implemented in Reaktoro (e.g., Peng-Robinson) are currently behaving in a way that it can model liquid, gaseous, and supercritical states. The defined gaseous phase should behave exclusively as gaseous or supercritical. However, for the conditions above, Peng-Robinson understands that the gas phase, with only water vapor, is more stable at liquid conditions (which is true). In the end, the equilibrium calculation ends up with all water from the aqueous phase migrating to the "gaseous phase" which is in fact in liquid state.

As a temporary workaround, try to define the gaseous phase as:

```py
gas = rkt.GaseousPhase("H2O(g) Ntg(g)")
```

For your vapor pressure calculation, however, you don't need to have a gaseous phase:

```py
import reaktoro as rkt
import math

db = rkt.PhreeqcDatabase("pitzer.dat")

# ph adjusted and charge neutralized 
feed_comp = {
    "H+": 1.9558980691081993e-08,
    "H2O": 55.50929780738274,
    "CO3-2": 4.905954807916131e-05,
    "CO2": 0.00026983065832876047,
    "Ca+2": 0.0037427017316233346,
    "Cl-": 0.048207460278976974,
    "HCO3-": 0.007875574268951177,
    "SO4-2": 1.0524665304507918e-05,
    "HSO4-": 5.9841513528777e-12,
    "Na+": 0.04871732291003011,
    "OH-": 5.429515778317798e-07,
}

solution = rkt.AqueousPhase(rkt.speciate(" ".join(list(feed_comp.keys()))))
solution.set(rkt.ActivityModelPitzer())

system = rkt.ChemicalSystem(db, solution)

state = rkt.ChemicalState(system)

for ion, values in feed_comp.items():
    state.set(ion, values, "mol")

state.temperature(293.15, "kelvin")
state.pressure(1, "bar")

rkt.equilibrate(state)

props = state.props()

T = props.temperature()
P = props.pressure()

μH2Oa  = props.speciesChemicalPotential("H2O")
μ0H2Og = db.species("H2O(g)").props(T, P).G0

RT = rkt.universalGasConstant * props.temperature()

pH2Og = math.exp((μH2Oa - μ0H2Og)/RT) * 1e5  # convert from bar to Pa

print("expected vapor pressure", 2351.213686124685, "Pa")
print("reaktoro vapor pressure", pH2Og, "Pa")
```

Assuming H2O is in equilibrium with H2O(g):

$$\mu_{\mathrm{H_{2}O}}=\mu_{\mathrm{H_{2}O(g)}},$$

where:

$$\mu_{\mathrm{H_{2}O}}=\mu_{\mathrm{H_{2}O(g)}}^{\circ}+RT\ln a_{\mathrm{H_{2}O(g)}}.$$

If we approximate:

$$a_{\mathrm{H_{2}O(g)}}\approx P_{\mathrm{H_{2}O(g)}}^{\mathrm{(bar)}}$$

we then arrive at the equation used above:

$$P_{\mathrm{H_{2}O(g)}}^{\mathrm{(bar)}}\approx\exp\left(\frac{\mu_{\mathrm{H_{2}O}}-\mu_{\mathrm{H_{2}O(g)}}^{\circ}}{RT}\right).$$

The output I got was:

```text
expected vapor pressure 2351.213686124685 Pa
reaktoro vapor pressure 2320.444058406082 Pa
```

**Reply by avdudchenko** [reply]
Thanks for the explanation and examples.   

One question: Would this be considered a "user error" or a "bug" in the EOS implementation? (I am almost erring on side of user error)

The reason I ask is to figure out if this should potentially be made into a doc/tutorial (or if we should expect changes to reaktoro in the future to address this (as well as have people who are invested help fix the issue)? This sort of calculation is quite common for water treatment/environmental engineering applications where we are looking at desalination or evaporation processes that operate at ambient or under "vacuum" conditions, such as mechanical vapor compression, membrane distillation, evaporation ponds etc, so either having clear example or work on a bug fix would be ideal. 

Also thank you for making the direct calculation example without gas phase, its great and I will implement it in our workflow for now, as we don't quite need the gas phase in general (it complicates things for process simulation).  

**For those reading in the future,** a key note on the "workaround" is a requirement for epsilon in solver; in rkt.equlibrate the default is 1e-16, so the workaround produced incorrect estimates of vapor pressure, but by using solver configured with an epsilon of 1e-32 resulted in an expected solution. This is due to the equilibrium mol amount of H2O(g) being ~2e-18. 

The example code with the workaround is below:
```py
import reaktoro as rkt

db = rkt.PhreeqcDatabase("pitzer.dat")

feed_comp = {
    "H+": 1.9558980691081993e-08,
    "H2O": 55.50929780738274,
    "CO3-2": 4.905954807916131e-05,
    "CO2": 0.00026983065832876047,
    "Ca+2": 0.0037427017316233346,
    "Cl-": 0.048207460278976974,
    "HCO3-": 0.007875574268951177,
    "SO4-2": 1.0524665304507918e-05,
    "HSO4-": 5.9841513528777e-12,
    "Na+": 0.04871732291003011,
    "OH-": 5.429515778317798e-07,
}

"""solve with added gas phase"""
solution = rkt.AqueousPhase(rkt.speciate(" ".join(list(feed_comp.keys()))))
solution.set(rkt.ActivityModelPitzer())
gas = rkt.GaseousPhase("H2O(g) Ntg(g)")
gas.set(rkt.ActivityModelPengRobinsonPhreeqc())
system_with_gas = rkt.ChemicalSystem(db, solution, gas)
state_with_gas = rkt.ChemicalState(system_with_gas)

for ion, values in feed_comp.items():
    state_with_gas.set(ion, values, "mol")
state_with_gas.temperature(293.15, "kelvin")
state_with_gas.pressure(1, "bar")

rkt.equilibrate(state_with_gas)
specs_with_gas = rkt.EquilibriumSpecs(system_with_gas)
specs_with_gas.temperature()
specs_with_gas.pressure()
solver = rkt.EquilibriumSolver(specs_with_gas)
conditions_with_gas = rkt.EquilibriumConditions(specs_with_gas)
conditions_with_gas.temperature(293.15, "kelvin")
conditions_with_gas.pressure(1, "bar")
solver_options = rkt.EquilibriumOptions()
solver_options.epsilon = 1e-32
solver.setOptions(solver_options)
result = solver.solve(state_with_gas, conditions_with_gas)
vapor_pressure = state_with_gas.props().speciesActivity("H2O(g)") * 101325
```

Output 
```expected vapor pressure 2351.213686124685  (Pa), reaktoro vapor pressure 2351.19 (Pa)```

And composition
```
+-----------------+------------+------+
| Property        |      Value | Unit |
+-----------------+------------+------+
| Temperature     |   293.1500 |    K |
| Pressure        |     1.0000 |  bar |
| Charge:         | 4.7321e-12 |  mol |
| Element Amount: |            |      |
| :: H            | 1.1103e+02 |  mol |
| :: C            | 8.1945e-03 |  mol |
| :: O            | 5.5534e+01 |  mol |
| :: Na           | 4.8717e-02 |  mol |
| :: Ntg          | 1.0000e-16 |  mol |
| :: S            | 1.0525e-05 |  mol |
| :: Cl           | 4.8207e-02 |  mol |
| :: Ca           | 3.7427e-03 |  mol |
| Species Amount: |            |      |
| :: H+           | 1.9559e-08 |  mol |
| :: H2O          | 5.5509e+01 |  mol |
| :: CO3-2        | 4.9060e-05 |  mol |
| :: CO2          | 2.6983e-04 |  mol |
| :: Ca+2         | 3.7427e-03 |  mol |
| :: Cl-          | 4.8207e-02 |  mol |
| :: HCO3-        | 7.8756e-03 |  mol |
| :: SO4-2        | 1.0525e-05 |  mol |
| :: HSO4-        | 5.9842e-12 |  mol |
| :: Na+          | 4.8717e-02 |  mol |
| :: OH-          | 5.4295e-07 |  mol |
| :: H2O(g)       | 2.3839e-18 |  mol |
| :: Ntg(g)       | 1.0000e-16 |  mol |
+-----------------+------------+------+
```

**Comment by allanleal** [comment]
Hi @avdudchenko - great that you could revise the original script and get the result correct as well, by tweaking the `epsilon` parameter. 

As to the current cubic EOS behavior in Reaktoro (modeling liquid, gas, and supercritical states), this was designed to account for the fact that some gaseous phases, under certain temperature and pressure conditions, can indeed exist as a liquid. For instance, a CO2-H2O phase at temperatures below 31.4°C and under high pressure will be in a liquid state. The cubic EOS implementation in Reaktoro correctly captures this behavior.

However, when dealing with a pure H2O gas phase, a conflict arises with the H2O in the aqueous phase. The cubic EOS for the H2O vapor phase will identify its most stable state as liquid and cause the equilibrium calculation to transfer all H2O from the aqueous phase to the pure H2O phase modeled with the cubic EOS.

To address this issue, adding an inert gas species, for now, to the vapor phase is an effective workaround. I have some ideas on how to resolve this problem more permanently. We can expect a future improvement that will allow for the modeling of a gas phase containing only H2O without encountering this issue.
